### DISCUSSION TITLE
What is the way to chain methods with different signatures describing a business process?

### DISCUSSION THREAD

**Discussion by Andras-Csanyi** [vavr-io/vavr#2825] [Q&A] [answered]
Hi all,

I am new in the functional programming world, so please be gentle... :D And I believe my Java knowledge is a bit rusty too.

I have the following process I'd like to describe with chained methods returning `Either`.

The process receives a context object containing a `WhateverInput` and `correlationId` and processes these in the following steps:

- validation of the WhateverInput
- mapping the WhateverInput to WhateverEntity
- saving the WhateverEntity in a storage
- mapping the returned WhateverEntity (having Id at this point) to WhateverResult

The `correlationId` is there to mark which process the operation belongs to if the process fails.

What I tried is `map` and `flatMap`, but both of them expects a function where the input is the Either's Right and an Either as a return value. However my method signatures look like below due to that I don't want to put the correlation id into a field.

```
public class VavrPractice {

    public static void handle(Context context) {

        var result = Either.right(context)
            .map(validateInput(context.getWhateverEntityInput(), context.getCorrelationId())) // shows error
    }

    private static Either<ErrorResult, WhateverEntityInput> validateInput(WhateverEntityInput whateverEntityInput, UUID correlationId) {}

    private static Either<ErrorResult, WhateverEntity> mapToWhateverEntity(WhateverEntityInput whateverEntityInput, UUID correlationId) {}

    private static Either<ErrorResult, WhateverEntity> saveToDatabase(WhateverEntityInput whateverEntityInput, UUID correlationId) {}

    private static Either<ErrorResult, WhateverEntityResult> mapToWhateverEntityResult(WhateverEntity whateverEntity, UUID correlationId) {}
}
```

How to do this in a nice, best/good practices way? What I can think of is creating a `Context` object adding to it both the `WhateverInput` and `correlationId` and this `Context` will be the `Right` side of the `Either` and moving this object between methods. But, it feels a bit clunky, heavy.

I study functional programming in both Java and C# in the same time and the example above can be done in C# using `Bind` and I am looking for an equivalent for that or the solution how the `Bind`, what it seems like `map` or `flatMap`, can be done.

Thanks the answers in advance.

**Comment by Andras-Csanyi** [comment]
I posted this to [Stackoverlow](https://stackoverflow.com/questions/78978224/what-is-the-way-to-chain-methods-with-different-signatures-describing-a-business).

**Comment by EarthCitizen** [comment, marked as answer]
@Andras-Csanyi Where you are using `.map()`, you need to use `.flatMap()` (bind). `.map()` is for only transforming what is inside the `Either.Right`.

Also, `.map()` and `.flatMap()` expect a function as a parameter, so you need to do this:

```java
.flatMap(__ -> validateInput(context.getWhateverEntityInput(), context.getCorrelationId()))
```

I am using `__` here because even though `.flatMap()` will provide `context` to its function parameter, this is not needed in this case because `context` is already in scope from the `handle()` method.

**Comment by EarthCitizen** [comment]
If you don't want to create a bunch of classes that represent transitional stages between your methods, you could simply use a `Tuple`, and it can grow and shrink as needed as each stage of your process is completed.
